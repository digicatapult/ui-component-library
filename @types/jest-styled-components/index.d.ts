declare module 'jest-styled-components/serializer'
