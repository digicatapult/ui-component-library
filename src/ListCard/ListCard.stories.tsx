import React from 'react'
import { ListCard } from '../index.js'
import { Story } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import styled from 'styled-components'

export default {
  title: 'Components/ListCard',
  component: ListCard,
  argTypes: {
    title: {
      control: {
        type: 'text',
      },
    },
    subtitle: {
      control: {
        type: 'text',
      },
    },
    orientation: {
      control: {
        type: 'select',
        options: ['left', 'right'],
      },
    },
    flashColor: {
      control: {
        type: 'text',
      },
    },
    background: {
      control: {
        type: 'text',
      },
    },
  },
}

const DefaultStoryTemplate: Story = (args) => {
  return (
    <ListCard
      title={args.title}
      subtitle={args.subtitle}
      orientation={args.orientation}
      flashColor={args.flashColor}
      background={args.background}
      height="67px"
      width="467px"
      onClick={action('click')}
    />
  )
}

export const Default = DefaultStoryTemplate.bind({})
Default.args = {
  title: 'Title',
  subtitle:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat',
  orientation: 'left',
  flashColor: 'darkblue',
  background: 'lightyellow',
}
export const HII = DefaultStoryTemplate.bind({})
HII.args = {
  title: 'Name',
  subtitle: 'Name of lead partner',
  orientation: 'left',
  flashColor: '#80CC72',
  background: '#DCE5E7',
}
export const Simple = DefaultStoryTemplate.bind({})
Simple.args = {
  title: 'Name',
  subtitle: undefined,
  orientation: undefined,
  flashColor: undefined,
  background: undefined,
}

export const ListCardMultiple: Story = () => {
  const colors = ['#00ffff', '#822c68', '#f6d6ff', '#0000ca', '#0a002d']
  const arr = Array(20)
    .fill(null)
    .map((_, i) => i)

  return (
    <ListWrapper>
      {arr.map((i) => (
        <ListCard
          title={`Card ${i}`}
          subtitle="Name of lead partner"
          orientation="left"
          flashColor={colors[i % colors.length]}
          background="#DCE5E7"
          height="5em"
          width="100%"
          onClick={action('click')}
        />
      ))}
    </ListWrapper>
  )
}

const ListWrapper = styled.div`
  display: grid;
  gap: 5px;
  font-size: 1.5em;
  width: 467px;
  height: 500px;
  overflow: scroll;
`
