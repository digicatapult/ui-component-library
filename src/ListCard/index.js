import ListCard from './ListCard.tsx'
export default ListCard
