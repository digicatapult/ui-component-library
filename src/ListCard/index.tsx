import React, { useRef } from 'react'
import styled from 'styled-components'
import { useId } from 'react-id-generator'

export interface ListCardProps {
  title: string
  subtitle?: string
  orientation?: 'left' | 'right'
  flashColor?: string
  background?: string
  width?: string
  height?: string
  onClick: (title: string) => {}
}

interface WrapperProps extends React.DOMAttributes<HTMLButtonElement> {
  orientation: 'left' | 'right'
  flashColor: string
  background: string
  width: string
  height: string
}

const ListCard: React.FC<ListCardProps> = ({
  flashColor = '#e0e0e0',
  background = '#f0f0f0',
  title,
  subtitle = '',
  orientation = 'left',
  width = '100%',
  height = '100%',
  onClick,
}) => {
  const [id] = useId()

  return (
    <Wrapper
      orientation={orientation}
      flashColor={flashColor}
      background={background}
      width={width}
      height={height}
      id={id}
      onClick={() => onClick(title)}
    >
      <label htmlFor={id}>{title}</label>
      {subtitle && <span>{subtitle}</span>}
    </Wrapper>
  )
}

const Wrapper = styled.button<WrapperProps>`
  border: 0;
  font: inherit;
  text-align: ${({ orientation }) => orientation};
  cursor: pointer;

  position: relative;
  width: ${({ width }) => width};
  height: ${({ height }) => height};
  padding-left: calc(1em * 41 / (16 * 1.2));
  overflow: hidden;

  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 0.5em;

  background: ${({ background }) => background};

  border-top-left-radius: ${({ orientation }) =>
    orientation === 'left' ? '0.75em' : '0'};
  border-top-right-radius: ${({ orientation }) =>
    orientation === 'right' ? '0.75em' : '0'};
  border-bottom-right-radius: ${({ orientation }) =>
    orientation === 'right' ? '0.75em' : '0'};
  border-bottom-left-radius: ${({ orientation }) =>
    orientation === 'left' ? '0.75em' : '0'};

  > * {
    display: block;
    padding: 0;
    margin: 0;
    width: 90%;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    cursor: pointer;
  }

  > label {
    font-size: 1.2em;
    font-weight: bolder;
  }

  > span {
    font-size: 1em;
  }

  ::before {
    content: '';
    position: absolute;
    top: 0;
    bottom: 0;
    left: ${({ orientation }) => (orientation === 'right' ? 'unset' : '0')};
    right: ${({ orientation }) => (orientation === 'left' ? 'unset' : '0')};

    width: ${({ orientation }) =>
      orientation === 'left' || orientation === 'right' ? '1em' : 'unset'};
    background: ${({ flashColor }) => flashColor};
  }

  :focus-visible {
    outline: 1px ${({ flashColor }) => flashColor} solid;
  }
`

export default ListCard
