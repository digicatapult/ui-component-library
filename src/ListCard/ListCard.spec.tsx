import React from 'react'
import renderer, { act } from 'react-test-renderer'
import 'jest-styled-components'

import ListCard from './index.js'

describe('Test', () => {
  test('default', () => {
    const clickHandler = jest.fn(() => {})
    const tree = renderer
      .create(<ListCard title="LIST CARD" onClick={clickHandler} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('hii', () => {
    const clickHandler = jest.fn(() => {})
    const tree = renderer
      .create(
        <ListCard
          title="Name"
          subtitle="Name of lead partner"
          orientation="left"
          flashColor="#80CC72"
          background="#DCE5E7"
          height="67px"
          width="467px"
          onClick={clickHandler}
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('orientation: right', () => {
    const clickHandler = jest.fn(() => {})
    const tree = renderer
      .create(
        <ListCard
          title="Name"
          subtitle="Name of lead partner"
          orientation="right"
          flashColor="#80CC72"
          background="#DCE5E7"
          height="67px"
          width="467px"
          onClick={clickHandler}
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('click', async () => {
    const clickHandler = jest.fn(() => {})
    const instance = renderer.create(
      <ListCard
        title="Test"
        subtitle="Name of lead partner"
        orientation="left"
        flashColor="#80CC72"
        background="#DCE5E7"
        height="67px"
        width="467px"
        onClick={clickHandler}
      />,
    )

    console.log(instance.root.findByType('button').props)
    await act(() => {
      instance.root.findByType('button').props.onClick({})
    })

    expect(clickHandler.mock.calls).toHaveLength(1)

    const call: string[] = clickHandler.mock.calls[0]
    expect(call[0]).toBe('Test')
  })
})
