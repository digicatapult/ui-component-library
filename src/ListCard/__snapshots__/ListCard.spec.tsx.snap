// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test default 1`] = `
.c0 {
  background: none;
  border: 0;
  font: inherit;
  text-align: left;
  cursor: pointer;
  position: relative;
  width: 100%;
  height: 100%;
  padding-left: calc(1em * 41 / (16 * 1.2));
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 0.5em;
  border-top-left-radius: 0.75em;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border-bottom-left-radius: 0.75em;
}

.c0 >* {
  display: block;
  padding: 0;
  margin: 0;
  width: 90%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  cursor: pointer;
}

.c0 >label {
  font-size: 1.2em;
  font-weight: bolder;
}

.c0 >span {
  font-size: 1em;
}

.c0::before {
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: unset;
  width: 1em;
  background: #e0e0e0;
}

.c0:focus-visible {
  outline: 1px #e0e0e0 solid;
}

<button
  background="none"
  className="c0"
  flashColor="#e0e0e0"
  height="100%"
  icon={false}
  id="id1"
  onClick={[Function]}
  orientation="left"
  width="100%"
>
  <label
    htmlFor="id1"
  >
    LIST CARD
  </label>
</button>
`;

exports[`Test hii 1`] = `
.c0 {
  background: #DCE5E7;
  border: 0;
  font: inherit;
  text-align: left;
  cursor: pointer;
  position: relative;
  width: 467px;
  height: 67px;
  padding-left: calc(1em * 41 / (16 * 1.2));
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 0.5em;
  border-top-left-radius: 0.75em;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border-bottom-left-radius: 0.75em;
}

.c0 >* {
  display: block;
  padding: 0;
  margin: 0;
  width: 90%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  cursor: pointer;
}

.c0 >label {
  font-size: 1.2em;
  font-weight: bolder;
}

.c0 >span {
  font-size: 1em;
}

.c0::before {
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: unset;
  width: 1em;
  background: #80CC72;
}

.c0:focus-visible {
  outline: 1px #80CC72 solid;
}

<button
  background="#DCE5E7"
  className="c0"
  flashColor="#80CC72"
  height="67px"
  icon={false}
  id="id2"
  onClick={[Function]}
  orientation="left"
  width="467px"
>
  <label
    htmlFor="id2"
  >
    Name
  </label>
  <span>
    Name of lead partner
  </span>
</button>
`;

exports[`Test orientation: right 1`] = `
.c0 {
  background: #DCE5E7;
  border: 0;
  font: inherit;
  text-align: right;
  cursor: pointer;
  position: relative;
  width: 467px;
  height: 67px;
  padding-left: calc(1em * 41 / (16 * 1.2));
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 0.5em;
  border-top-left-radius: 0;
  border-top-right-radius: 0.75em;
  border-bottom-right-radius: 0.75em;
  border-bottom-left-radius: 0;
}

.c0 >* {
  display: block;
  padding: 0;
  margin: 0;
  width: 90%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  cursor: pointer;
}

.c0 >label {
  font-size: 1.2em;
  font-weight: bolder;
}

.c0 >span {
  font-size: 1em;
}

.c0::before {
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: unset;
  right: 0;
  width: 1em;
  background: #80CC72;
}

.c0:focus-visible {
  outline: 1px #80CC72 solid;
}

<button
  background="#DCE5E7"
  className="c0"
  flashColor="#80CC72"
  height="67px"
  icon={false}
  id="id3"
  onClick={[Function]}
  orientation="right"
  width="467px"
>
  <label
    htmlFor="id3"
  >
    Name
  </label>
  <span>
    Name of lead partner
  </span>
</button>
`;
